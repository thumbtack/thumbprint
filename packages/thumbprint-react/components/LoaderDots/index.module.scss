@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

.root {
    display: flex;
    justify-content: center;
    list-style: none;
    padding: 0;
    margin: 0;
}

.dot {
    border-radius: $tp-border-radius__full;
    animation: tp-loader-animation-1 1.3s infinite ease-in;
    opacity: 0;
    -moz-transform: rotate(0.1deg); // Fixes small rendering bug in FF;
}

.dot:nth-child(2) {
    animation-name: tp-loader-animation-2;
}

.dot:nth-child(3) {
    animation-name: tp-loader-animation-3;
}

.dotSizeMedium {
    width: 16px;
    height: 16px;
}

.dotSizeMedium:nth-child(2) {
    margin-left: 10px;
    margin-right: 10px;
}

.dotSizeSmall {
    width: 8px;
    height: 8px;
}

.dotSizeSmall:nth-child(2) {
    margin-left: 5px;
    margin-right: 5px;
}

.dotThemeBrand {
    background-color: $tp-color__blue;
}

.dotThemeInverse {
    background-color: $tp-color__white;
}

.dotThemeMuted {
    background-color: $tp-color__black-300;
}

/**
 * Text only visible to screenreaders.
 * http://webaim.org/techniques/css/invisiblecontent/
 * TODO: Potentially abstract into a reusable component.
 */
.hiddenText {
    position: absolute;
    left: -10000px;
    top: auto;
    width: 1px;
    height: 1px;
    overflow: hidden;
}

@keyframes tp-loader-animation-1 {
    0%,
    70%,
    100% {
        opacity: 0.2;
    }
    30% {
        opacity: 1;
    }
}

@keyframes tp-loader-animation-2 {
    0%,
    15%,
    85%,
    100% {
        opacity: 0.2;
    }
    45% {
        opacity: 1;
    }
}

@keyframes tp-loader-animation-3 {
    0%,
    30%,
    100% {
        opacity: 0.2;
    }
    60% {
        opacity: 1;
    }
}
