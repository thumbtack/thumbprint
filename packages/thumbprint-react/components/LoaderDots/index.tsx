import React from 'react';
import classNames from 'classnames';
import styles from './index.module.scss';

interface PropTypes {
    /**
     * Text that describes the current status and is only visible to
     * screenreaders.
     */
    assistiveText?: string;
    /**
     * Controls the size of the dots.
     */
    size?: 'small' | 'medium';
    /**
     * Changes the dot colors.
     */
    theme?: 'brand' | 'inverse' | 'muted';
}

export default function LoaderDots({
    assistiveText = 'Loading',
    size = 'medium',
    theme = 'brand',
}: PropTypes): JSX.Element {
    const dotClasses = classNames({
        [styles.dot]: true,
        [styles.dotThemeBrand]: theme === 'brand',
        [styles.dotThemeInverse]: theme === 'inverse',
        [styles.dotThemeMuted]: theme === 'muted',
        [styles.dotSizeSmall]: size === 'small',
        [styles.dotSizeMedium]: size === 'medium',
    });

    return (
        <ul className={styles.root} role="status">
            <li className={dotClasses} />
            <li className={dotClasses} />
            <li className={dotClasses} />
            <li className={styles.hiddenText}>{assistiveText}</li>
        </ul>
    );
}
