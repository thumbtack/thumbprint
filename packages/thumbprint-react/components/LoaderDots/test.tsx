import React from 'react';
import { mount } from 'enzyme';
import LoaderDots from './index';

test('has proper default attributes', () => {
    const wrapper = mount(<LoaderDots />);

    expect(wrapper.find('.dotThemeBrand')).toHaveLength(3);
    expect(wrapper.find('.dotSizeMedium')).toHaveLength(3);
    expect(wrapper.find('.root').prop('role')).toBe('status');
    expect(wrapper.find('.hiddenText').text()).toBe('Loading');
    expect(wrapper).toMatchSnapshot();
});

test('adds proper accessibility attributes', () => {
    const wrapper = mount(<LoaderDots assistiveText="Goose" />);
    expect(wrapper.find('.root').prop('role')).toBe('status');
    expect(wrapper.find('.hiddenText').text()).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('sets size attribute', () => {
    const wrapper = mount(<LoaderDots size="small" />);
    expect(wrapper.find('.dotSizeSmall')).toHaveLength(3);
    expect(wrapper).toMatchSnapshot();
});

test('sets theme attribute', () => {
    const wrapperInverse = mount(<LoaderDots theme="inverse" />);
    expect(wrapperInverse.find('.dotThemeInverse')).toHaveLength(3);
    expect(wrapperInverse).toMatchSnapshot();

    const wrapperMuted = mount(<LoaderDots theme="muted" />);
    expect(wrapperMuted.find('.dotThemeMuted')).toHaveLength(3);
    expect(wrapperMuted).toMatchSnapshot();
});
