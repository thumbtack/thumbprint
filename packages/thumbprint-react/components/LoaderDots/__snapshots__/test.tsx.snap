// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds proper accessibility attributes 1`] = `
<LoaderDots
  assistiveText="Goose"
>
  <ul
    className="root"
    role="status"
  >
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="hiddenText"
    >
      Goose
    </li>
  </ul>
</LoaderDots>
`;

exports[`has proper default attributes 1`] = `
<LoaderDots>
  <ul
    className="root"
    role="status"
  >
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="dot dotThemeBrand dotSizeMedium"
    />
    <li
      className="hiddenText"
    >
      Loading
    </li>
  </ul>
</LoaderDots>
`;

exports[`sets size attribute 1`] = `
<LoaderDots
  size="small"
>
  <ul
    className="root"
    role="status"
  >
    <li
      className="dot dotThemeBrand dotSizeSmall"
    />
    <li
      className="dot dotThemeBrand dotSizeSmall"
    />
    <li
      className="dot dotThemeBrand dotSizeSmall"
    />
    <li
      className="hiddenText"
    >
      Loading
    </li>
  </ul>
</LoaderDots>
`;

exports[`sets theme attribute 1`] = `
<LoaderDots
  theme="inverse"
>
  <ul
    className="root"
    role="status"
  >
    <li
      className="dot dotThemeInverse dotSizeMedium"
    />
    <li
      className="dot dotThemeInverse dotSizeMedium"
    />
    <li
      className="dot dotThemeInverse dotSizeMedium"
    />
    <li
      className="hiddenText"
    >
      Loading
    </li>
  </ul>
</LoaderDots>
`;

exports[`sets theme attribute 2`] = `
<LoaderDots
  theme="muted"
>
  <ul
    className="root"
    role="status"
  >
    <li
      className="dot dotThemeMuted dotSizeMedium"
    />
    <li
      className="dot dotThemeMuted dotSizeMedium"
    />
    <li
      className="dot dotThemeMuted dotSizeMedium"
    />
    <li
      className="hiddenText"
    >
      Loading
    </li>
  </ul>
</LoaderDots>
`;
